### PR TITLE
fact gathering, mounts, fixes for single proc code and tests

### PR DESCRIPTION
**Pull request by bcoca**
Instead of 2 code paths, use dummy pool
Moved pool selection to function for easier unit test
updated integration and added unit test

##### ISSUE TYPE

- Bugfix Pull Request

### CODE REVIEW

File: lib/ansible/module_utils/facts/hardware/linux.py [hunk 1]
```diff
@@ -27,6 +27,7 @@
 
 from multiprocessing import cpu_count
 from multiprocessing.pool import ThreadPool
+from multiprocessing.dummy.pool import Pool as DummyPool
```
Comment (mattclay) [inline lib/ansible/module_utils/facts/hardware/linux.py]:
```suggestion
from multiprocessing.dummy import Pool as DummyPool
```

File: test/units/module_utils/facts/hardware/linux/test_get_pool.py [hunk 2]
```diff
@@ -17,7 +17,7 @@ def test_get_pool():
 
 def test_get_dummy_pool():
 
-    with mock.patch('multiprocessing.pool.ThreadPool', new=mock.MagicMock()) as TPool:
-        TPool.__new__ = mock.MagicMock(side_effect=PermissionError('pooof'))
+    with mock.patch('multiprocessing.pool.ThreadPool.__new__', new=mock.MagicMock()) as TPool:
+        TPool = mock.MagicMock(side_effect=PermissionError('pooof'))
```
Comment (mattclay) [inline test/units/module_utils/facts/hardware/linux/test_get_pool.py]: The mock can simplified further:

```suggestion
    with mock.patch.object(multiprocessing.pool.ThreadPool, '__init__', side_effect=PermissionError('poof')):
```

**Comment by mattclay** [comment]
@bcoca It turns out the built-in dummy pool also uses `/dev/shm`, even though it doesn't need to, since it's only using threads. Thankfully there is a fairly simple solution for that:

```python
import threading

from multiprocessing.pool import ThreadPool, Pool
from multiprocessing.context import BaseContext

class DummyContext(BaseContext):
    def Lock(self):
        return threading.Lock()


class DummyPool(ThreadPool):
    def __init__(self, processes=None, initializer=None, initargs=()):
        Pool.__init__(self, processes, initializer, initargs, context=DummyContext())
```

This `DummyPool` can be used instead of `multiprocessing.dummy.Pool`.

File: lib/ansible/module_utils/common/process.py [hunk 3]
```diff
@@ -4,11 +4,36 @@
...
+    def Lock(self):
+        return threading.Lock()
+
+
+class DummyPool(ThreadPool):
+    def __init__(self, processes=None, initializer=None, initargs=()):
+        Pool.__init__(self, processes, initializer, initargs, context=DummyContext())
+
+
+def get_thread_pool(size, warn=None):
```
Comment (mattclay) [inline lib/ansible/module_utils/common/process.py]: Until we have a use case for making this utility code, we should probably keep it within the fact gathering code that needs it.

File: lib/ansible/module_utils/common/process.py [hunk 4]
```diff
@@ -4,11 +4,36 @@
...
 
 
+class DummyContext(BaseContext):
+    def Lock(self):
+        return threading.Lock()
+
+
+class DummyPool(ThreadPool):
+    def __init__(self, processes=None, initializer=None, initargs=()):
+        Pool.__init__(self, processes, initializer, initargs, context=DummyContext())
```
Comment (mattclay) [inline lib/ansible/module_utils/common/process.py]:
```suggestion
        Pool.__init__(self, processes, initializer, initargs, context=DummyContext())  # pylint: disable=non-parent-init-called
```

File: lib/ansible/module_utils/common/process.py (hunk 3 above)
Comment (bcoca) [reply lib/ansible/module_utils/common/process.py]: if it were the single function, I would agree, but if we add this kind of required scafolding, it makes little sense to keep there.

Comment (mattclay) [reply lib/ansible/module_utils/common/process.py]: If we end up needing it in other modules later, we can move it into module_utils. Even then, unless there's a need to make it part of our public API, it should be private.

File: lib/ansible/module_utils/facts/hardware/linux.py [hunk 5]
```diff
@@ -612,66 +602,44 @@ def get_mount_facts(self):
...
-            else:
-                # use multiproc pool, handle results below
-                results[mount]['extra'] = pool.apply_async(self.get_mount_info, (mount, device, uuids))
+            results[mount]['extra'] = pool.apply_async(self.get_mount_info, (mount, device, uuids))
 
-        if pool is None:
-            # serial processing, just assing results
-            mounts.append(results[mount]['info'])
-        else:
             pool.close()  # done with spawing new workers, start gc
```
Comment (mattclay) [inline lib/ansible/module_utils/facts/hardware/linux.py]: The pool needs to be closed outside the for loop:

```suggestion
        pool.close()  # done with spawning new workers, start gc
```

File: lib/ansible/plugins/action/copy.py [hunk 6]
```diff
@@ -310,13 +310,11 @@ def _copy_file(self, source_full, source_rel, content, content_tempfile,
...
-            # world readable.  Access to the directory itself is controlled via fixup_perms2() as
-            # part of executing the module. Check that umask with scp/sftp/piped doesn't cause
-            # a problem before acting on this idea. (This idea would save a round-trip)
-            # fix file permissions when the copy is done as a different user
+            # File source permissions are attempted to be preserved via sftp/scp or local
+            # but they might fail and/or use other connections or transfer methods
+            # So we ensure that the remote is readable by remote user (could be moved to copy module)
             if remote_path:
-                self._fixup_perms2((self._connection._shell.tmpdir, remote_path))
+                self._fixup_perms2((self._connection._shell.tmpdir, remote_path), execute=False)
```
Comment (mattclay) [inline lib/ansible/plugins/action/copy.py]: It looks like these changes are unrelated to fact gathering. Were they supposed to be on another PR?

Comment (bcoca) [reply lib/ansible/plugins/action/copy.py]: yep

File: lib/ansible/module_utils/facts/hardware/linux.py [hunk 7]
```diff
@@ -21,28 +21,22 @@
...
 import sys
 import time
 
-from multiprocessing import cpu_count
-from multiprocessing.pool import ThreadPool
-
-from ansible.module_utils.common.text.converters import to_text
+from ansible.module_utils._internal._concurrent import _futures
 from ansible.module_utils.common.locale import get_best_parsable_locale
+from ansible.module_utils.common.text.converters import to_text
```
Comment (mattclay) [inline lib/ansible/module_utils/facts/hardware/linux.py:29]: The original import above can be kept instead of moving it here.

File: lib/ansible/module_utils/facts/hardware/linux.py [hunk 8]
```diff
@@ -21,28 +21,22 @@
...
 from ansible.module_utils.common.text.formatters import bytes_to_human
 from ansible.module_utils.facts.hardware.base import Hardware, HardwareCollector
 from ansible.module_utils.facts.utils import get_file_content, get_file_lines, get_mount_size
 from ansible.module_utils.six import iteritems
 
+
```
Comment (mattclay) [inline lib/ansible/module_utils/facts/hardware/linux.py]: This extra line isn't needed.

File: lib/ansible/module_utils/facts/hardware/linux.py (hunk 7 above)
Comment (bcoca) [reply lib/ansible/module_utils/facts/hardware/linux.py:29]: its easier to find/update stuff if they are ordered
